R/derive_vars_trt_end.R

```diff
@@ -0,0 +1,152 @@
+#'
+#' @author Stefan Bundfuss, Teckla Akinyi
+#'
+#' @return The input dataset with start date variable added e.g. `TRTEDTM`
+#'
+#' @export
+#'
+#' @keywords adsl timing derivation
+#'
+#' @examples
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("ex")
+#' data("dm")
+#'
+#'dm %>%
+#'  derive_vars_trt_end(dataset_ex = ex,
+#'                      new_var=TRTEDTM,
+#'                      dtc=EXSTDTC,
+#'                      date_imputation = "first",
+#'                      time_imputation="first",
+#'                      min_dates = NULL,
+#'                      max_dates = NULL,
+#'                      order = vars(EXSTDTC,EXSEQ),
+#'                      subject_keys=vars(STUDYID,USUBJID),
+#'                      flag_imputation = "AUTO",
+#'                      filter_ex = EXDOSE>0 ,
+#'                      ord_filter="LAST")
+#'   select(USUBJID, TRTEDTM)
+derive_vars_trt_end <- function(dataset,
+                                  dataset_ex,
+                                  filter_ex = NULL,
+                                  subject_keys,
+                                  new_var,
+                                  dtc,
+                                  date_imputation,
+                                  time_imputation,
+                                  min_dates = NULL,
+                                  max_dates = NULL,
+                                  flag_imputation = "AUTO",
+                                  order,
+                                  ord_filter) {
+
+  new_var <- assert_symbol(enquo(new_var))
+  dtc <- assert_symbol(enquo(dtc))
+  assert_data_frame(dataset, subject_keys)
+  assert_data_frame(dataset_ex, required_vars = quo_c(subject_keys, order))
+  assert_character_scalar(ord_filter, values = c("first", "last"), case_sensitive = F)
+
+  filter_ex <- assert_filter_cond(enquo(filter_ex), optional = TRUE)
+  filter_ex <- enquo(filter_ex)
+
+  assert_character_scalar(flag_imputation, values = c("auto", "both", "date", "time", "none"),
+                          case_sensitive = FALSE)
+
+  #Issue warning that new_var should be DTM
+  if (str_detect(rlang::quo_name(new_var), "DTM", negate = T)) {
+    msg <- sprintf(
+      "This function derives the treatment datetime, new_var must end in DTM.")
+    warn(msg)
+  }
+
+  add <- dataset_ex %>%
+    mutate(imputed_dtc = convert_dtc_to_dtm(!!dtc,
+                                            date_imputation = date_imputation,
+                                            time_imputation = time_imputation,
+                                            min_dates = min_dates,
+                                            max_dates = max_dates)
+    )
+
+  #filter EX data based on study requirements
+  if (quo_is_null(filter_ex)) {
+    add
+    msg <- sprintf(
+      "Input EX dataset has no been filtered. Check study requirement if ex needs to be filtered.")  # nolint
+    warn(msg)
+  } else {
+    add <- add %>% filter(!!filter_ex)
+  }
```

Comment:
I would remove the warning because if `filter_ex` is null, it was specified by the user. See my comment above regarding the default value.

I would apply the filter before the imputation by calling `filter_if()`. This is more efficient.